src/test/java/com/puppycrawl/tools/checkstyle/checks/imports/UnusedImportsCheckTest.java

```diff
@@ -36,6 +45,41 @@ protected String getPackageLocation() {
     }
 
     @Test
+    @SuppressWarnings("unchecked")
+    public void testReferencedStateIsClearedOnBeginTree() throws Exception {
+        final UnusedImportsCheck check = new UnusedImportsCheck();
+        check.setProcessJavadoc(false);
+        final File input = new File(getPath("InputUnusedImportCheckClearState.java"));
+        final DetailAST importAst = TestUtils.parseFile(input).getNextSibling();
+        final DetailAST classDef = importAst.getNextSibling().getNextSibling();
+        final Optional<DetailAST> identAst = TestUtils.findTokenInAstByPredicate(classDef,
+            token -> token.getType() == TokenTypes.IDENT);
+
+        assertTrue("IDENT is not present in import tree", identAst.isPresent());
+        check.beginTree(null);
+        check.visitToken(classDef);
+        check.visitToken(identAst.get());
+        check.beginTree(null);
+        final Field field = check.getClass().getDeclaredField("referenced");
+        field.setAccessible(true);
+
+        assertTrue("Stateful field is not cleared after beginTree",
+            ((Collection<String>) field.get(check)).isEmpty());
```

Comment:
I need good reason why this can not be tested by `verify`.
Usage of reflection is ok , only when you prove that it is impossible to cover is logic is working properly.
All should be described in comment above this method.